Sources/TSCBasic/Process.swift

```diff
@@ -648,22 +655,26 @@ public final class Process {
                                 pending = readResult
                             }
                         }
+                        group.leave()
                     } else if let stdoutResult = (pendingLock.withLock { pending }) {
                         // TODO: this is more of an error
                         self?.stateLock.withLock {
                             self?.state = .outputReady(stdout: stdoutResult, stderr: .success([]))
                         }
+                        group.leave()
                     }
                 }
             } else {
                 pendingLock.withLock {
                     pending = .success([])  // no stderr in this case
                 }
             }
+            
             // first set state then start reading threads
             self.stateLock.withLock {
-                self.state = .readingOutputThread(stdout: stdoutThread, stderr: stderrThread)
+                self.state = .readingOutputPipe(sync: group)
```

Comment:
we should probably rename this to something else than readingOutputPipe if reusing it for the thread bases scenario as well, how about `readingOutput`

Reply:
Sounds good.